scripts/train_lm.py

```diff
@@ -146,12 +146,39 @@ def WriteMetaparameters(metaparameters, ngram_order, num_train_sets, out_file):
 
 def FormatMetaparameters(metaparameters):
     out = []
+    # get the maximum length of decimals of metaparameters
+    decimal_length = 0
     for param in metaparameters:
-        x = '{:.3f}'.format(param)
-        if x == '0.00':
-            x = '{:.3g}'.format(param)
-        out.append(x)
-
+        temp_length = len(str(param))
+        if decimal_length < temp_length:
+            decimal_length = temp_length
+    decimal_length -= 2
+    
+    temp_original = metaparameters[0]
+    temp = round(temp_original, 3)
+    for param in metaparameters[1:]:
+        round_index = 3
+        x_original = param
+        x = round(x_original, round_index)
+        while(temp == x and round_index <= decimal_length):
```

Comment:
Also, there is no need to have special code that sets things to 0.001 or
0.999 if they would be rounded to 0 or 1.  Just keep adding more digits
until it's not exactly zero or 1.


On Tue, Aug 16, 2016 at 6:13 PM, Ke Li <notifications@github.com> wrote:

> In scripts/train_lm.py
> <https://github.com/danpovey/pocolm/pull/52#discussion_r75049323>:
>
> > -        if x == '0.00':
> > -            x = '{:.3g}'.format(param)
> > -        out.append(x)
> > -
> > +        temp_length = len(str(param))
> > +        if decimal_length < temp_length:
> > +            decimal_length = temp_length
> > +    decimal_length -= 2
> > +
> > +    temp_original = metaparameters[0]
> > +    temp = round(temp_original, 3)
> > +    for param in metaparameters[1:]:
> > +        round_index = 3
> > +        x_original = param
> > +        x = round(x_original, round_index)
> > +        while(temp == x and round_index <= decimal_length):
>
> Hi Dan,
> Ok, I'll change the variables' names.
> Actually I think my code can deal with the two problems base on following
> your suggestion as "by e.g. rounding them to 3 decimals but then if you
> detect that you're about to repeat a
> value or produce an exact 0 or 1, add decimal places until that is no
> longer the case". I assumed the collisions only occur in the way that the
> repeating values occur as a pair of two connective name numbers after
> rounding. I guess maybe this assumption is wrong.
> So you want me to rewrite the code to handle cases like more than two
> values are same or the same values are not connective? Or I just need to
> reformat the code (e.g. add helper functions) to make it more clear?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/danpovey/pocolm/pull/52/files/3bb42c2f1362ffd85d6a4cf0b1fe1a50ab2fe9e1#r75049323>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu64nmjKEoQAWW4y25Qv59MHlb_kBks5qgmBVgaJpZM4Jl6Bt>
> .
>